refs #12147 - fixing dhcp test to work properly with rails 4

unorthodoxgeek opened the pull request:
In rails 4, the #reload method doesn't reset instance variables on the
reloaded instance, so we fetch it again as a new item. That said, the
reload method does work correctly for associations, just to be
consistent. Updated test to work for both rails 3 and 4.

dLobatog commented on lib/tasks/trends.rake:
```diff
@@ -13,7 +13,7 @@ namespace :trends do
 
     trends.each do |trend_id|
-      puts "Working on trend_id #{trend_id}, #{(current_record += 1)} of #{trends_count}"
+      puts "Working on trend_id #{trend_id}, #{(current_record += 1)} of #{trends_count}" unless Rails.env.test?
```
This one is ok if it's not included - it was just a way to make tests output less verbose.

unorthodoxgeek replied:
I'm for this, it's annoying, and should be fixed, if we're at it, let's do this as well

dLobatog commented:
@unorthodoxgeek In general I know why these changes were needed and they are necessary in Rails 4. We should try to merge this knowing why and what exactly are we fixing. A good way to do so and get things in faster, and make reviewer's life easy (which makes developer life easier as well) would be to:

1 - Find if other parts of the code need more fixes like this (if any).
2 - Submit a PR with the changes, explaining in the commit description why exactly a fix is necessary, why is it failing on Rails 4, in what way, and how we fix it. I know https://github.com/theforeman/foreman/pull/2358 didn't do it and it was a mistake as we don't know exactly what was that commit for. I tried to amend this mistake and submit things in a more sensible way from there onwards.

For instance in this PR, the `.reload` fix has :zero: to do with the :factory: Factory fix, and they have nothing to do with the ensure_no_cycle fix. I'm fine merging some of these individually right away, like factories or seeds_test (when there's a good description!), but some others like `ensure_no_cycle` require a bit more testing and reviewing - I wrote that code and I feel like it needs closer inspection :smile: 

All in all what I'm trying to say is that if this PR is split into a few others with good descriptions each PR will be a lot less risk and so much easier to understand in the end it gets merged faster. Does that sound good? You could keep the same issue number even, and modify the title and description for one of the issues.

unorthodoxgeek committed:
refs #12147 - fixing dhcp test to work properly with rails 4

In rails 4, the #reload method doesn't reset instance variables on the
reloaded instance, so we fetch it again as a new item. That said, the
reload method does work correctly for associations, just to be
consistent. Updated test to work for both rails 3 and 4.

unorthodoxgeek commented:
@eLobato - seperated the PR into 4 different PRs, each touching only one thing, with descriptive commit messages.

dLobatog commented:
Merged as 5c8deb3a721d395b65a45a48b2c43d52e89a268f, thanks @unorthodoxgeek!
